roxie/ccd/ccdfile.cpp

```diff
@@ -587,12 +602,175 @@ static void appendRemoteLocations(IPartDescriptor *pdesc, StringArray &locations
+
+struct CacheInfoEntry
+{
+    union
+    {
+        struct
+        {
+            bool diskCache : 1;   // false means it's in the jhtree cache, true means it's only in OS disk cache
+            offset_t page: 39;
+            unsigned file: 24;
+        } b;
+        offset_t u;
+    };
+
+    inline CacheInfoEntry() { u = 0; }
+    inline CacheInfoEntry(unsigned _file, offset_t _pos, bool _diskCache)
+    {
+        b.file = _file;
+        b.page = _pos >> pageBits;
+        b.diskCache = _diskCache;
+    }
+    inline bool operator < ( const CacheInfoEntry &l) const { return u < l.u; }
+    inline bool operator <= ( const CacheInfoEntry &l) const { return u <= l.u; }
+    inline void operator++ () { b.page++; }
+
+    static constexpr unsigned pageBits = 13;  // 8k 'pages'
+};
+
+class CacheReportingBuffer : public ICacheInfoRecorder
+{
+    CacheInfoEntry *recentReads = nullptr;
+    std::atomic<unsigned> recentReadHead = {0};
+    unsigned recentReadSize;
+
+public:
+
+    CacheReportingBuffer(offset_t trackSize)
+    {
+        recentReadSize = trackSize >> CacheInfoEntry::pageBits;
+        if (traceLevel)
+            DBGLOG("Creating CacheReportingBuffer with %d elements", recentReadSize);
+        assertex(recentReadSize);
+        recentReads = new CacheInfoEntry[recentReadSize];
+        memset(recentReads, 0, recentReadSize * sizeof(CacheInfoEntry));
+        recentReadHead = 0;
+    }
+    CacheReportingBuffer(const CacheReportingBuffer &from)
+    {
+        recentReadSize = from.recentReadSize;
+        recentReads = new CacheInfoEntry[recentReadSize];
+        memcpy(recentReads, from.recentReads, recentReadSize * sizeof(CacheInfoEntry));
+        recentReadHead = from.recentReadHead.load(std::memory_order_relaxed);
+    }
+
+    ~CacheReportingBuffer()
+    {
+        delete [] recentReads;
+    }
+
+    void clear()
+    {
+        recentReadHead = 0;
+    }
+
+    void noteRead(unsigned fileIdx, offset_t pos, unsigned len, bool diskCache)
+    {
+        if (recentReads && len)
+        {
+            CacheInfoEntry start(fileIdx, pos, diskCache);
+            CacheInfoEntry end(fileIdx, pos+len, diskCache);
+            for(;start <= end; ++start)
+            {
+                recentReads[recentReadHead++ % recentReadSize] = start;
+            }
+        }
+    }
+
+    void sortAndDedup()
+    {
+        unsigned sortSize;
+        if (recentReadHead > recentReadSize)
+            sortSize = recentReadSize;
+        else
+            sortSize = recentReadHead;
+        std::sort(recentReads, recentReads + sortSize);
+        CacheInfoEntry lastPos(-1,-1,false);
+        unsigned dest = 0;
+        for (unsigned idx = 0; idx < sortSize; idx++)
+        {
+            CacheInfoEntry pos = recentReads[idx];
+            if (pos.b.file != lastPos.b.file || pos.b.page != lastPos.b.page)   // Ignore inNodeCache bit when deduping
```

Comment:
I don't know how likely they are. And I'm not sure I understand the suggestion.

Reply:
The reason I remove them rather than skipping them on output is that I had the idea I might move the deduped array back to the active buffer.

This code (when reporting) is not time-critical